File: src/agentgate/domain/result.py
"""Detailed evaluation results."""

from __future__ import annotations

from enum import StrEnum
from typing import Any, Literal
from uuid import uuid4

from pydantic import Field, field_serializer, field_validator, model_validator

from .base import DomainModel, freeze_json
from .evaluation import Dimension, JudgeEvidence, Kind, MethodRef, Severity


class Outcome(StrEnum):
    PASS = "pass"
    FAIL = "fail"
    REVIEW = "review"
    NOT_APPLICABLE = "not_applicable"
    ERROR = "error"


class FailureStage(StrEnum):
    ROUTING = "routing"
    TOOL_SELECTION = "tool_selection"
    TOOL_ARGUMENTS = "tool_arguments"
    TOOL_EXECUTION = "tool_execution"
    FINAL_STATE = "final_state"
    FINAL_OUTPUT = "final_output"


class FailureObservation(DomainModel):
    stage: FailureStage
    observed_at_sequence: int = Field(ge=0)
    span_id: str | None = None


class Evidence(DomainModel):
    trace_id: str
    span_ids: tuple[str, ...] = ()
    description: str
    document_refs: tuple[str, ...] = ()


class EvaluationErrorEvidence(DomainModel):
    category: Literal["crash", "timeout", "invalid_output"]
    exception_type: str
    message: str
    retryable: bool = False
    reference: str | None = None


class CheckResult(DomainModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    name: str
    turn_id: str | None = None
    expectation_id: str | None = None
    outcome: Outcome
    score: float | None = Field(default=None, ge=0, le=1)
    reason: str
    expected: Any = None
    actual: Any = None
    actual_missing: bool = False
    methods: tuple[MethodRef, ...] = ()
    evidence: tuple[Evidence, ...] = ()
    failure_observation: FailureObservation | None = None

    @field_validator("expected", "actual", mode="before")
    @classmethod
    def freeze_values(cls, value: Any) -> Any:
        return freeze_json(value)

    @field_serializer("expected", "actual")
    def serialize_values(self, value: Any) -> Any:
        from .base import thaw_json
        return thaw_json(value)

    @model_validator(mode="after")
    def validate_outcome_fields(self) -> CheckResult:
        if self.outcome == Outcome.FAIL and self.failure_observation is None:
            raise ValueError("failed checks require failure_observation")
        if self.outcome != Outcome.FAIL and self.failure_observation is not None:
            raise ValueError("only failed checks may have failure_observation")
        if self.outcome in (Outcome.NOT_APPLICABLE, Outcome.ERROR) and self.score is not None:
            raise ValueError("not-applicable/error checks cannot have a score")
        if self.outcome in (Outcome.PASS, Outcome.FAIL, Outcome.REVIEW) and self.score is None:
            raise ValueError("measured checks require a score")
        return self


class Result(DomainModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    run_id: str
    case_id: str
    evaluator_id: str
    evaluator_name: str
    evaluator_version: str
    evaluator_kind: Kind
    dimension: Dimension
    metric: str
    severity: Severity
    outcome: Outcome
    score: float | None = Field(default=None, ge=0, le=1)
    reason: str
    checks: tuple[CheckResult, ...] = ()
    methods: tuple[MethodRef, ...] = ()
    evidence: tuple[Evidence, ...] = ()
    judge_evidence: JudgeEvidence | None = None
    error_evidence: EvaluationErrorEvidence | None = None
    primary_failure_step: FailureStage | None = None
    trace_revision: int | None = Field(default=None, ge=1)
    trace_content_sha256: str | None = None

    @model_validator(mode="after")
    def validate_result(self) -> Result:
        if self.outcome in (Outcome.NOT_APPLICABLE, Outcome.ERROR) and self.score is not None:
            raise ValueError("not-applicable/error results cannot have a score")
        if self.outcome in (Outcome.PASS, Outcome.FAIL, Outcome.REVIEW) and self.score is None:
            raise ValueError("measured results require a score")
        if self.outcome == Outcome.FAIL:
            if self.primary_failure_step is None:
                raise ValueError("failed results require primary_failure_step")
            if not any(item.outcome == Outcome.FAIL for item in self.checks):
                raise ValueError("failed results require at least one failed CheckResult")
        elif self.primary_failure_step is not None:
            raise ValueError("only failed results may have primary_failure_step")
        if self.outcome == Outcome.ERROR:
            if self.error_evidence is None or self.primary_failure_step is not None:
                raise ValueError("error results require error evidence and no failure stage")
        elif self.error_evidence is not None:
            raise ValueError("only error results may carry error evidence")
        return self

File: web/src/types/result.ts
// Result 域类型
import type { Run } from './target'

export interface Evidence {
  trace_id: string
  span_ids: string[]
  description: string
}

export type Outcome = 'pass' | 'fail' | 'review' | 'not_applicable' | 'error'

/** 评估方法引用：对比算子（equals/contains_all/...），来自后端 CheckResult.methods */
export interface MethodRef {
  operator: string
  operator_version: string
  condition_kind?: string | null
}

export interface CheckResult {
  id: string
  name: string
  turn_id: string | null
  expectation_id: string | null
  outcome: Outcome
  score: number | null
  reason: string
  expected: unknown
  actual: unknown
  actual_missing: boolean
  methods?: MethodRef[]
  evidence: Evidence[]
}

export interface Result {
  case_id: string
  evaluator_id: string
  evaluator_name: string
  evaluator_kind: string
  dimension: string
  metric: string
  severity: 'standard' | 'blocking'
  outcome: Outcome
  score: number | null
  reason: string
  primary_failure_step?: string
  evidence: Evidence[]
  checks: CheckResult[]
}

export interface Gate {
  outcome: 'pass' | 'fail'
  passed: number
  failed: number
  reviewed: number
  not_applicable: number
  errors: number
  score: number | null
  threshold: number
  reason: string
}

export interface Metric {
  key: string
  label: string
  level: 'overall' | 'kind' | 'dimension' | 'metric'
  score: number | null
  passed: number
  failed: number
  reviewed: number
  not_applicable: number
  errors: number
  applicable: number
  total: number
  incomplete: boolean
}

export interface Report {
  run: Run
  results: Result[]
  gate: Gate
  metrics: Metric[]
}

export type ComparisonStatus = 'improved' | 'regressed' | 'unchanged' | 'incomparable'

export interface RerunComparison {
  root_run_id: string
  parent_run_id: string
  rerun_run_id: string
  case_id: string
  case_name: string
  before_target_version: string
  after_target_version: string
  overall: ComparisonStatus | 'mixed'
  counts: Record<ComparisonStatus, number>
  evaluators: {
    evaluator_id: string
    evaluator_name: string
    status: ComparisonStatus
    before: { outcome: Outcome; score: number | null; reason: string } | null
    after: { outcome: Outcome; score: number | null; reason: string } | null
  }[]
}
